daemon/cmd/fqdn.go

```diff
@@ -525,6 +527,43 @@ func (d *Daemon) notifyOnDNSMsg(lookupTime time.Time, ep *endpoint.Endpoint, epI
 	record.Log()
 
 	if msg.Response && msg.Rcode == dns.RcodeSuccess && len(responseIPs) > 0 {
+		// Create a critical section especially for when multiple DNS requests
+		// are in-flight for the same name (i.e. cilium.io).
+		//
+		// It is possible for the following race to occurr:
+		//
+		//              G1                                    G2
+		//
+		// T0 --> NotifyOnDNSMsg()               NotifyOnDNSMsg()            <-- T0
+		//
+		// T1 --> UpdateGenerateDNS()            UpdateGenerateDNS()         <-- T1
+		//
+		// T2 ----> mutex.Lock()                 +---------------------------+
+		//                                       |No identities need updating|
+		// T3 ----> mutex.Unlock()               +---------------------------+
+		//
+		// T4 --> UpsertGeneratedIdentities()    UpsertGeneratedIdentities() <-- T4
+		//
+		// T5 ---->  Upsert()                    DNS released back to pod    <-- T5
+		//                                                    |
+		// T6 --> DNS released back to pod                    |
+		//              |                                     |
+		//              |                                     |
+		//              v                                     v
+		//       Traffic flows fine                   Leads to policy drop
+		//
+		// Note how G2 releases the DNS msg back to the pod at T5 because
+		// UpdateGenerateDNS() was a no-op. It's a no-op because G1 had executed
+		// UpdateGenerateDNS() first at T1 and performed the necessary identity
+		// allocation for the response IPs. Due to G1 performing all the work
+		// first, G2 executes T4 also as a no-op and releases the msg back to the
+		// pod at T5 before G1 would at T6.
+		mutexes := d.getMutexesForResponseIPs(responseIPs)
+		for _, m := range mutexes {
```

Comment:
It'd be nice to be able to measure this mutex acqusition time, in order to assist tweaking the new value. For instance if it's seconds then we should probably be bumping the mutex count.

At the same time, I imagine these defaults should be fine for most users. So, should we add additional spans to the current `stat` and generate all those extra datapoints in metrics that most user shouldn't care about? Or is it enough to just wrap this loop with a timestamp check and just raise a warning if it takes longer than N seconds?

Reply:
By default I'm somewhat inclined to go for the latter. Then the question is what that limit should be before emitting a warning. We could also make _that_ configurable. Probably we could go for something relatively loose like 500ms to begin and tweak further from there. In most cases it should be drastically lower, but in worst case maybe this includes some level of policy generation + ipcache generation time that goes on a bit long(?).

Reply:
I've added a new flag as mentioned.